### PR TITLE
Cleanup/clarification of PIO behavior with >32 pins

### PR DESCRIPTION
**Pull request by kilograham**

### CODE REVIEW

File: src/rp2_common/hardware_pio/include/hardware/pio.h [hunk 1]
```diff
@@ -69,6 +69,31 @@
...
+ * \if rp2040_specific
+ * On RP2040, pin numbers may always be specified from 0-31
+ * \endif
+ *
+ * \if rp2350_specific
+ * On RP2350A, pin numbers may always be specified from 0-31.
+ *
+ * On RP2350B, there are 48 pins but each PIO instance can only address 32 pins (the PIO
+ * instance either addresses pins 0-31 or 16-47 based on \ref pio_set_gpio_base). The
+ * `pio_sm_` methods that directly affect the hardware always take _real_ pin numbers in the full range, however:
```
Comment (peterharperuk) [inline src/rp2_common/hardware_pio/include/hardware/pio.h:83]: Is it worth saying that sm_ methods also take _real_ pin numbers?

Comment (peterharperuk) [reply src/rp2_common/hardware_pio/include/hardware/pio.h:83]: ah - you've said that below - ignore

**Review by peterharperuk — APPROVED** [review]
Looks good to me, and fixes the uart_rx example!

**Review by peterharperuk — CHANGES_REQUESTED** [review]
ah - I think pio_sm_set_pindirs_with_mask64 is shifting twice. It shifts by pio_get_gpio_base then calls pio_sm_set_pindirs_with_mask_internal which shifts again. Removing the shift from pio_sm_set_pindirs_with_mask_internal fixes this. uart_tx can be used to test this.

File: src/rp2_common/hardware_pio/pio.c [hunk 2]
```diff
@@ -260,21 +269,35 @@ void pio_sm_set_pins_with_mask64(PIO pio, uint sm, uint64_t pinvals, uint64_t pi
...
     }
     pio->sm[sm].pinctrl = pinctrl_saved;
     pio->sm[sm].execctrl = execctrl_saved;
 }
 
-void pio_sm_set_pindirs_with_mask(PIO pio, uint sm, uint32_t pindirs, uint32_t pin_mask) {
-    pio_sm_set_pindirs_with_mask64(pio, sm, pindirs, pin_mask);
+#ifndef pio_sm_set_pins_with_mask_internal
+void pio_sm_set_pins_with_mask(PIO pio, uint sm, uint32_t pin_values, uint32_t pin_mask) {
+#if PICO_PIO_USE_GPIO_BASE
```
Comment (peterharperuk) [inline src/rp2_common/hardware_pio/pio.c:281]: These lines need to be removed?

**Review by peterharperuk — APPROVED** [review]
Seems to work nicely

File: src/rp2_common/hardware_pio/pio.c [hunk 3]
```diff
@@ -240,18 +240,27 @@ void pio_sm_set_pins64(PIO pio, uint sm, uint64_t pins) {
     pio->sm[sm].execctrl = execctrl_saved;
 }
 
-void pio_sm_set_pins_with_mask(PIO pio, uint sm, uint32_t pinvals, uint32_t pin_mask) {
-    pio_sm_set_pins_with_mask64(pio, sm, pinvals, pin_mask);
+#ifndef pio_sm_set_pins_internal
```
Comment (lurch) [inline src/rp2_common/hardware_pio/pio.c:243]: I'm sure this must be a stupid question, but why is line 243 doing
```
#ifndef pio_sm_set_pins_internal
```
when line 221 says
```
void pio_sm_set_pins_internal(PIO pio, uint sm, uint32_t pins) {
```
?

**Pull request merged**